Use dictionaries instead of tuples for linkifiers.

abhijeetbodas2001 opened the pull request:
https://chat.zulip.org/#narrow/stream/49-development-help/topic/Nullable.20element.20in.20TupleType

This adds new event type `realm_linkifier` which uses dictionaries instead of tuples.


**Testing plan:** <!-- How have you tested? -->
This PR updates the webapp frontend to handle the new-type events, but I also tested manually **before** making the frontend changes to verify that the old style events are still sent as before.

<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

showell commented on zerver/lib/event_schema.py:
```diff
@@ -734,9 +734,28 @@ def check_realm_export(
+
+# This is a legacy event type to ensure backwards compatibility
+# for old clients. Newer clients should handle only the
+# "realm_linkifiers" event above.
```
Does `realm_filter_type` get used any more after your changes?  We should keep it, since we still need to support old clients.  But we probably won't test it in `test_events`.  Instead, we'll probably want you to introduce some test in `test_linkifiers.py` or similar that simulates an old client connecting to us, and we'll want to make sure the payload adheres to this schema.  I'm not exactly sure where that test should go, actually, and there might be an existing test that we just modify.

showell commented:
This looks very clean and thorough!  One quick question is where do we test with the old schema (i.e. `realm_filter_type` via `check_realm_filters` now in the backend tests?

abhijeetbodas2001 commented:
The last push was just a change to the commit body so as to mention that we send both the old and the new event types.

timabbott commented on zerver/lib/markdown/__init__.py:
```diff
@@ -1762,8 +1762,8 @@ def prepare_realm_pattern(source: str) -> str:
 # using the provided format string to construct the URL.
-class RealmFilterPattern(markdown.inlinepatterns.Pattern):
-    """ Applied a given realm filter to the input """
+class RealmLinkifierPattern(markdown.inlinepatterns.Pattern):
```
LinkifierPattern would be a better name, I think, especially since our plans include making Stream-level ones possible as well?

abhijeetbodas2001 replied:
Yep! Renamed to `LinkifierPattern`.

timabbott commented on zerver/management/commands/realm_linkifiers.py:
```diff
@@ -2,24 +2,24 @@
-Example: ./manage.py realm_filters --realm=zulip --op=remove '#(?P<id>[0-9]{2,8})'
-Example: ./manage.py realm_filters --realm=zulip --op=show
+Example: ./manage.py realm_linkifier --realm=zulip --op=remove '#(?P<id>[0-9]{2,8})'
+Example: ./manage.py realm_linkifier --realm=zulip --op=show
```
Let's rename this to `manage.py edit_linkifiers`.

abhijeetbodas2001 replied:
Done, renamed to `management/commands/edit_linkifiers.py`

timabbott commented on zerver/management/commands/realm_linkifiers.py:
```diff
@@ -41,24 +41,24 @@ def handle(self, *args: Any, **options: str) -> None:
         pattern = options["pattern"]
         if not pattern:
-            self.print_help("./manage.py", "realm_filters")
+            self.print_help("./manage.py", "realm_linkifier")
```
I think there's probably a prep commit you could do to make this access its own command name via sys.args[1] or something rather than hardcoding?  Not sure how easy this is with Django's management command system, but that'd be a good sweep to do.

abhijeetbodas2001 replied:
I added a prep commit which does that sweep, and yet another one before it as a small correction 😄

timabbott commented on zerver/models.py:
```diff
@@ -874,61 +874,61 @@ def __str__(self) -> str:
 
 
-def all_realm_filters() -> Dict[int, List[Tuple[str, str, int]]]:
+def linkifiers_from_all_realms() -> Dict[int, List[Tuple[str, str, int]]]:
```
I think I like `all_linkifiers_for_installation`  better as a name. -- we sometimes use "installation" as the term for an entire Zulip server in the backend.

abhijeetbodas2001 replied:
Done, renamed to `all_linkifiers_for_installation`.

timabbott commented on zerver/models.py:
```diff
@@ -874,61 +874,61 @@ def __str__(self) -> str:
-    global per_request_realm_filters_cache
-    per_request_realm_filters_cache = {}
+    global per_request_linkifiers_cache
+    per_request_linkifiers_cache = {}
```
Definitely don't try to do this change in this commit, but I think we can actually rename `RealmFilter` to `Linkifier` in `models.py` with a small Django configuration option to set the database table name to still be `zerver_realmfilter`.

abhijeetbodas2001 replied:
I haven't put any of the following in this PR, but just writing my findings here...

There's a `db_table` meta property which we can specify for the table name. So a diff like this-

```diff
-class RealmFilter(models.Model):
+class Linkifier(models.Model):
     """Realm-specific regular expressions to automatically linkify certain
     strings inside the Markdown processor.  See "Custom filters" in the settings UI.
     """
@@ -869,6 +869,7 @@ class RealmFilter(models.Model):
     url_format_string: str = models.TextField(validators=[URLValidator(), filter_format_validator])

     class Meta:
+        db_table = "zerver_realmfilter"
         unique_together = ("realm", "pattern")
```
Results in a migration like so-
```python
    operations = [
        migrations.RenameModel(
            old_name='RealmFilter',
            new_name='Linkifier',
        ),
        migrations.AlterModelTable(
            name='linkifier',
            table='zerver_realmfilter',
        ),
    ]

```

Is that what you have in mind?

Another question I have is, why not change the database table name? I tried running migrations both ways -- specifying the `db_name` to be the old `zerver_realmfilter` and not specifying it (so that it defaults to `zerver_linkifier`). There seems to be no data loss in either case.

timabbott commented on zerver/tests/test_realm_linkifiers.py:
```diff
@@ -69,13 +69,13 @@ def test_create(self) -> None:
 
         data["pattern"] = r"lp:(?P<id>[0-9]+)"
-        data["url_format_string"] = "https://realm.com/my_realm_filter/?sort=reverse&value=%(id)s"
+        data["url_format_string"] = "https://realm.com/my_linkifier/?sort=reverse&value=%(id)s"
```
Definitely don't this in the current commit, but ideally we should change these to use the IETF standard `example.com` as the main domain.  (Maybe `bugs.example.com/?sort...` for this launchpad example, and similar for the ones that are Jira or the like.

abhijeetbodas2001 replied:
I removed these specific changes from the commit. Many similar URL strings are also there in `markdown_test_cases.json`. Do those need to be changed as well?

timabbott commented:
I posted a hopefully complete batch of comments on the 2nd commit (the first was merged in #17625).  Thanks @abhijeetbodas2001, I'm very excited about this.

abhijeetbodas2001 commented:
@timabbott Thanks for the review! I addressed most of your comments, except the last couple ones. I didn't understand what you mean by "don't try to do this in this commit". Do you mean it should be extracted as a prep commit here, or a commit at the end of this PR, or as a different follow-up PR all together?

showell commented:
I would be inclined to do the change of moving things to `example.com` as a separate PR.  I'm not sure that's exactly what Tim intended, but I think the gist of his comment is that he thinks it's a worthwhile cleanup, just not in that particular commit.

abhijeetbodas2001 commented:
I've done some splitting of commits to try to make this more readable. The last commit is still huge (200+ lines), but a lot of it is `zulip.yaml`.

timabbott commented:
I got this test failure nondeterministically when running tests in parallel with the second commit included.  Haven't tried to debug.
```
======================================================================
ERROR: test_management_commands_show_help (zerver.tests.test_management_commands.TestCommandsCanStart) [<object object at 0x7f9afcc7aa80>] (management_command='check_redis')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3.6/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/usr/lib/python3.6/unittest/case.py", line 523, in subTest
    yield
  File "/home/tabbott/zulip/zerver/tests/test_management_commands.py", line 215, in test_management_commands_show_help
    call_command(command, "--help")
  File "/srv/zulip-py3-venv/lib/python3.6/site-packages/django/core/management/__init__.py", line 111, in call_command
    command = load_command_class(app_name, command_name)
  File "/srv/zulip-py3-venv/lib/python3.6/site-packages/django/core/management/__init__.py", line 37, in load_command_class
    module = import_module('%s.management.commands.%s' % (app_name, name))
  File "/srv/zulip-py3-venv/lib/python3.6/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 994, in _gcd_import
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 665, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 678, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/home/tabbott/zulip/zerver/management/commands/check_redis.py", line 13, in <module>
    class Command(BaseCommand):
  File "/home/tabbott/zulip/zerver/management/commands/check_redis.py", line 17, in Command
    Usage: ./manage.py [--trim] {sys.argv[1]}"""
IndexError: list index out of range

```

timabbott commented:
OK I merged most of the prep commits through 52a86d9604bfa3f9f304d766d10f0241c61759ea, since they're independent of that issue.

timabbott commented on frontend_tests/node_tests/markdown.js:
```diff
@@ -13,13 +13,19 @@ set_global("location", {
+    {
+        pattern: "ZGROUP_(?P<id>[0-9]{2,8}):(?P<zone>[0-9]{1,8})",
+        url_format: "https://zone_%(zone)s.zulip.net/ticket/%(id)s",
+    },
```
Should these examples have an ID field too?

abhijeetbodas2001 replied:
They aren't required for these tests -- the `id`s are used in only the settings page, but I'll add them here too for completeness.

abhijeetbodas2001 replied:
Done!

abhijeetbodas2001 replied:
I also did this treatment to a couple other instances in this file, so that these dictionaries are complete even if aren't testing all of the fields.

abhijeetbodas2001 commented:
@timabbott The main commit here is ready for another review. I've addressed https://github.com/zulip/zulip/pull/17596#discussion_r607424585.
I've removed the `sys.argv[1]` commit from this PR for now. I hadn't run into the failure above when I ran the test a couple times before pushing the commit earlier.
I don't have an idea right now about what causes it to break. I will try to investigate it a bit more thoroughly, and add that commit (after fixing it) back here once done!

abhijeetbodas2001 commented:
Is there a way to run only a few tests parallelly, without doing the whole of `test-backend` (which is pretty slow on my machine 😅)?

timabbott commented:
> Is there a way to run only a few tests parallelly, without doing the whole of test-backend (which is pretty slow on my machine sweat_smile)?

@abhijeetbodas2001 yeah, you can do `test-backend <suite1> <suite2> --parallel=2`.  It's all documented in `test-backend --help`.

timabbott commented:
This is great, merged, thanks @abhijeetbodas2001!  I made the following tweaks to documentation while merging:

```
diff --git a/templates/zerver/api/changelog.md b/templates/zerver/api/changelog.md
index f8d8dcccb3..a4fe251942 100644
--- a/templates/zerver/api/changelog.md
+++ b/templates/zerver/api/changelog.md
@@ -12,13 +12,20 @@ below features are supported.
 
 **Feature level 54**
 
-* `GET /realm/filters` is replaced by [`GET /realm/linkifiers`](/api/get-linkifiers)
-  which returns the data in a cleaner, dictionary format.
-* [`GET /events`](/api/get-events): Introduced new event type `realm_linkifiers`.
-  This will replace the `realm_filters` events in the future.
-* [`POST /register`](/api/register-queue): The response now also contains a
-  `realm_linkifiers` key, which is similar to the existing `realm_filters` key,
-  but uses dictionaries instead tuples.
+* `GET /realm/filters` has been removed and replace with [`GET
+  /realm/linkifiers`](/api/get-linkifiers) which returns the data in a
+  cleaner dictionary format.
+* [`GET /events`](/api/get-events): Introduced new event type
+  `realm_linkifiers`.  The previous `realm_filters` event type is
+  still supported for backwards compatibility, but will be removed in
+  a future release.
+* [`POST /register`](/api/register-queue): The response now supports a
+  `realm_linkifiers` event type, containing the same data as the
+  legacy `realm_filters` key, with a more extensible object
+  format. The previous `realm_filters` event type is still supported
+  for backwards compatibility, but will be removed in a future
+  release. The legacy `realm_filters` key is deprecated but remains
+  available for backwards compatibility.
 
 **Feature level 53**
 
diff --git a/zerver/openapi/zulip.yaml b/zerver/openapi/zulip.yaml
index d5c39141ee..8536ea34eb 100644
--- a/zerver/openapi/zulip.yaml
+++ b/zerver/openapi/zulip.yaml
@@ -2378,8 +2378,13 @@ paths:
                                 Processing this event is important to doing Markdown local echo
                                 correctly.
 
-                                **Changes**: New in Zulip 4.0 (feature level 54). This will eventually
-                                replace the `realm_filters` event type.
+                                **Changes**: New in Zulip 4.0 (feature level 54), replacing the
+                                previous `realm_filters` event type, which is still sent for
+                                backwards compatibility reasons.
+
+                                Clients should migrate to requesting and processing the
+                                `realm_linkifiers` event type when possible, since we plan to remove
+                                the legacy `realm_filters` logic entirely in a future release.
                               properties:
                                 id:
                                   $ref: "#/components/schemas/EventIdSchema"
@@ -2401,15 +2406,16 @@ paths:
                                         type: string
                                         description: |
                                           The string regex pattern which represents the pattern that
-                                          should be linkified on matching.
+                                          should be linkified by this linkifier.
                                       url_format:
                                         type: string
                                         description: |
-                                          The URL with which the pattern matching string should be linkified.
+                                          The URL format string to be used for linkifying matches.
+
                                       id:
                                         type: integer
                                         description: |
-                                          The ID of the realm linkifier.
+                                          The ID of the linkifier.
                               example:
                                 {
                                   "type": "realm_linkifiers",
@@ -2431,11 +2437,9 @@ paths:
                                 when the set of configured [linkifiers](/help/add-a-custom-linkifier)
                                 for the organization has changed.
 
-                                Processing this event is important to doing Markdown local echo
-                                correctly.
-
                                 **Changes**: Deprecated in Zulip 4.0 (feature level 54), replaced by
-                                the `realm_linkifiers` event type instead.
+                                the `realm_linkifiers` event type, which has a clearer name and format,
+                                instead.
                               properties:
                                 id:
                                   $ref: "#/components/schemas/EventIdSchema"
@@ -6556,8 +6560,9 @@ paths:
 
         `GET {{ api_url }}/v1/realm/linkifiers`
 
-        **Changes**: New in Zulip 4.0 (feature level 54). This replaces
-        the old `GET /realm/filters` endpoint completely.
+        **Changes**: New in Zulip 4.0 (feature level 54). On older versions,
+        a similar `GET /realm/filters` endpoint was available with each entry in
+        a `[pattern, url_format, id]` tuple format.
       responses:
         "200":
           description: Success.
@@ -6582,11 +6587,11 @@ paths:
                               type: string
                               description: |
                                 The string regex pattern which represents the pattern that
-                                should be linkified on matching.
+                                should be linkified by this linkifier.
                             url_format:
                               type: string
                               description: |
-                                The URL with which the pattern matching string should be linkified.
+                                The URL format string to be used for linkifying matches.
                             id:
                               type: integer
                               description: |
@@ -7085,7 +7090,9 @@ paths:
                           Array of objects where each object describes a single
                           [linkifier](/help/add-a-custom-linkifier).
 
-                          **Changes**: New in Zulip 4.0 (feature level 54).
+                          **Changes**: New in Zulip 4.0 (feature level 54). Clients can
+                          access these data on older server versions via the previous
+                          `realm_filters` key.
                         items:
                           type: object
                           additionalProperties: false
```

timabbott commented:
@abhijeetbodas2001 I think as a natural follow-up, we should change the `POST` endpoint for linkifiers as well.  That endpoint is only used by the webapp today, so we should be able to move it over without leaving a backwards-compatible option.

timabbott commented:
I've also opened issues for mobile and terminal to migrate to the new format. Thanks for doing this @abhijeetbodas2001! It's very nice to be on a path to eliminating these weird tuple quirks from our API documentation.

abhijeetbodas2001 commented:
Yay! I think that leaves just `muted_topic_type` events now? Is there a good enough reason to convert those too in the near future?

abhijeetbodas2001 commented:
> @abhijeetbodas2001 yeah, you can do test-backend <suite1> <suite2> --parallel=2. It's all documented in test-backend --help. It can also sometimes be helpful to run test-backend --reverse to look for order-dependent test failures.

Thanks. Pretty dumb of me to not look at `--help` before though 🤦‍♂️

abhijeetbodas2001 commented:
Let me compile a list of TODOs from here-

- [ ] `sys.argv[1]` sweep for management commands https://github.com/zulip/zulip/pull/17596#discussion_r594583246
- [ ] renaming the `RealmFilter` Django model to `Linkifier` https://github.com/zulip/zulip/pull/17596#discussion_r594585178
- [ ] Converting `realm/filters:POST` to `realm/linkifiers:POST`. The mobile repo [has code to call this enpoint](https://github.com/zulip/zulip-mobile/blob/master/src/api/realm/createRealmFilter.js), but it looks like that module isn't called from anywhere. Is it just dead code? Also, can we not rename all of these endpoint (`POST`, `DELETE`, and the upcoming `PATCH`)? None of the clients other than the webapp currently allow users to edit or change linkifiers correct?
- [ ] Using `example.com` everywhere instead of other real or dummy links https://github.com/zulip/zulip/pull/17596#discussion_r594586643
- [ ] Starting work on stream-specific linkifiers!

@timabbott I would like to tackle all of these. Is the priority order correct?
